### PR TITLE
RFC : Add DRM_PRIME Format Handling with KMS / DRM Atomic.

### PR DESCRIPTION
**Pull request by LongChair**
This commit allows to use the AV_PIX_FMT_DRM_PRIME newly introduced format in ffmpeg that allows decoders to provide an AVDRMFrameDescriptor struct.

That struct holds dmabuf fds and information allowing zerocopy rendering using KMS / DRM Atomic.

This has been tested on RockChip ROCK64 device.

I agree that my changes can be relicensed to LGPL 2.1 or later.

### CODE REVIEW

**Comment by wm4** [comment]
On a superficial look seems all good. I think there are some minor things I'd like to be changed before merge. Also hoping @rr- can comment.

Also CC @tobiasjakobi who possibly wanted to look into better DRM support including atomic, although this PR doesn't change vo_drm.c much.

**Comment by LongChair** [comment]
`vo_drm` could probably be adapted with the helpers from drm_atomic in more or less the same way :)

**Comment by tobiasjakobi** [comment]
Hello,

I guess this code doesn't have much intersection with mine, since it's more about efficiently passing hw-decoded buffers to the DRM subsystem. In my setup the decoding is done on the CPU. The resulting output should then be presented via DRM, but without relying on the CPU again (for steps like scaling or CSC).

The only thing I kinda noticed is that the new code mixed legacy DRM and atomic DRM. E.g. drmModeGetCrtc() and drmModeSetPlane() is stilled used.

**Comment by wm4** [comment]
Yes. But vo_drm.c could eventually also get native overlay support. I guess LongChair is for now only interested in using overlay with vo_gpu, because it solves the problem how subtitles should be rendered.

For mixed legacy/atomic, I think the idea is that the PR doesn't break drivers which do not provide atomic support yet.

**Comment by LongChair** [comment]
Yes vo_drm does all the copy from sourceframes to primary buffer. Frames could still be rendered to overlay and osd & subtitles to primary. Would make less copy and be more efficient :)

File: libmpv/opengl_cb.h [hunk 1]
```diff
@@ -163,9 +175,13 @@ extern "C" {
  *
  * glMPGetNativeDisplay("wl") should return a Wayland "struct wl_display *".
  *
- * glMPGetNativeDisplay("drm") should return a DRM FD casted to intptr_t (note
- * that a 0 FD is not supported - if this can happen in your case, you must
- * dup2() it to a non-0 FD).
+ * glMPGetNativeDisplay("drm") should return an array of void * containing
+ * 3 elements which are
```
Comment (rr-) [inline libmpv/opengl_cb.h]: Missing colon.

File: video/out/opengl/hwdec_drmprime_drm.c [hunk 2]
```diff
@@ -0,0 +1,255 @@
...
+#include "video/out/gpu/hwdec.h"
+#include "video/mp_image.h"
+
+#include "ra_gl.h"
+
+typedef struct drm_frame
+{
+    struct drm_prime_framebuffer fb;
+    struct mp_image *image; // associated mpv image
+
```
Comment (rr-) [inline video/out/opengl/hwdec_drmprime_drm.c]: Superfluous newline.

File: video/out/opengl/hwdec_drmprime_drm.c [hunk 3]
```diff
@@ -0,0 +1,255 @@
...
+#include "options/m_config.h"
+#include "video/out/drm_common.h"
+#include "video/out/drm_prime.h"
+#include "video/out/gpu/hwdec.h"
+#include "video/mp_image.h"
+
+#include "ra_gl.h"
+
+typedef struct drm_frame
+{
```
Comment (rr-) [inline video/out/opengl/hwdec_drmprime_drm.c]: K&R

File: video/out/opengl/hwdec_drmprime_drm.c [hunk 4]
```diff
@@ -0,0 +1,255 @@
...
+    struct priv *p = hw->priv;
+    drmModeCrtcPtr crtc;
+    double hratio, vratio, ratio;
+    int display_w = 0, display_h = 0;
+
+    // drm can allow to have a layer that has a different size from framebuffer
+    // we scale here the destination size to video mode
+    hratio = vratio = ratio = 1.0;
+
+    crtc = drmModeGetCrtc(p->ctx->fd, p->ctx->crtc->id);
```
Comment (rr-) [inline video/out/opengl/hwdec_drmprime_drm.c]: New allocation every frame, not sure if negligible

File: video/out/opengl/context_drm_egl.c [hunk 5]
```diff
@@ -241,17 +236,40 @@ static void acquire_vt(void *data)
...
+        p->kms->atomic_context->request = drmModeAtomicAlloc();
+        p->native_handles[2] = p->kms->atomic_context->request;
+    }
+}
+
 static void drm_egl_swap_buffers(struct ra_ctx *ctx)
 {
     struct priv *p = ctx->priv;
+    struct drm_atomic_context *atomic_ctx = p->kms->atomic_context;
+    int ret;
```
Comment (rr-) [inline video/out/opengl/context_drm_egl.c]: I'd add newline after var declarations

File: video/out/drm_prime.c [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+
+int drm_prime_create_framebuffer(struct mp_log *log, int fd, AVDRMFrameDescriptor *descriptor, int width, int height,
+                                  struct  drm_prime_framebuffer *framebuffer)
+{
+    AVDRMLayerDescriptor *layer = NULL;
+    uint32_t pitches[4], offsets[4], handles[4];
+    int ret, layer_fd;
+
+    if ((descriptor) && (descriptor->nb_layers)) {
+
```
Comment (rr-) [inline video/out/drm_prime.c]: Superfluous newline

File: video/out/drm_prime.c [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+#include "drm_prime.h"
+
+int drm_prime_create_framebuffer(struct mp_log *log, int fd, AVDRMFrameDescriptor *descriptor, int width, int height,
+                                  struct  drm_prime_framebuffer *framebuffer)
+{
+    AVDRMLayerDescriptor *layer = NULL;
+    uint32_t pitches[4], offsets[4], handles[4];
+    int ret, layer_fd;
+
+    if ((descriptor) && (descriptor->nb_layers)) {
```
Comment (rr-) [inline video/out/drm_prime.c]: Superfluous braces

File: video/out/drm_prime.c [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+
+        memset(framebuffer, 0, sizeof(*framebuffer));
+
+        for (int object=0; object < descriptor->nb_objects; object++) {
```
Comment (rr-) [inline video/out/drm_prime.c]: Spaces around `=`

File: video/out/drm_common.c [hunk 9]
```diff
@@ -260,6 +260,24 @@ struct kms *kms_create(struct mp_log *log, const char *connector_spec,
...
         goto err;
 
+    // Universal planes allows accessing all the planes (including primary)
+    if (drmSetClientCap(kms->fd, DRM_CLIENT_CAP_UNIVERSAL_PLANES, 1)) {
+        mp_err(log, "Failed to set Universal planes capability\n");
+    }
+
+    if (drmSetClientCap(kms->fd, DRM_CLIENT_CAP_ATOMIC, 1)) {
+        mp_verbose(log, "No DRM Atomic support found\n");
+    }
```
Comment (rr-) [inline video/out/drm_common.c]: Merge with the line below

File: video/out/drm_atomic.c [hunk 10]
```diff
@@ -0,0 +1,229 @@
...
+       }
+   }
+
+   return -EINVAL;
+}
+
+struct drm_object * drm_object_create(struct mp_log *log, int fd, uint32_t object_id, uint32_t type)
+{
+    struct drm_object *obj = NULL;
+    obj = malloc(sizeof(struct drm_object));
```
Comment (rr-) [inline video/out/drm_atomic.c]: Consider using talloc (applies to all `malloc` calls)

File: video/out/drm_atomic.c [hunk 11]
```diff
@@ -0,0 +1,229 @@
...
+        free(object);
+    }
+}
+
+void drm_object_print_info(struct mp_log *log, struct drm_object *object)
+{
+    mp_err(log, "Object ID = %d (type = %x) has %d properties\n",
+           object->id, object->type, object->props->count_props);
+
+    for (int i=0; i < object->props->count_props; i++)
```
Comment (rr-) [inline video/out/drm_atomic.c]: Spaces around `=`

File: video/out/drm_atomic.c [hunk 12]
```diff
@@ -0,0 +1,229 @@
...
+    memset(ctx, 0, sizeof(*ctx));
+
+    ctx->fd = fd;
+    ctx->crtc = drm_object_create(log, ctx->fd, crtc_id, DRM_MODE_OBJECT_CRTC);
+    if (!ctx->crtc) {
+        mp_err(log, "Failed to create CRTC object\n");
+        goto fail;
+    }
+
+    for (int i=0; i<res->count_crtcs; i++) {
```
Comment (rr-) [inline video/out/drm_atomic.c]: Spaces around operators

File: video/out/drm_atomic.c [hunk 13]
```diff
@@ -0,0 +1,229 @@
...
+        if (res->crtcs[i] == crtc_id) {
+            crtc_index = i;
+            break;
+        }
+    }
+
+    for (unsigned int j=0; j < plane_res->count_planes; j++) {
```
Comment (rr-) [inline video/out/drm_atomic.c]: Spaces around `=`

File: video/out/drm_atomic.c [hunk 14]
```diff
@@ -0,0 +1,229 @@
...
+
+        drmplane = drmModeGetPlane (ctx->fd, plane_res->planes[j]);
```
Comment (rr-) [inline video/out/drm_atomic.c]: Remove space before `(`

File: video/out/drm_atomic.c [hunk 15]
```diff
@@ -0,0 +1,229 @@
...
+        if (drmplane->possible_crtcs & (1 << crtc_index)) {
+
+            plane = drm_object_create(log, ctx->fd, drmplane->plane_id, DRM_MODE_OBJECT_PLANE);
```
Comment (rr-) [inline video/out/drm_atomic.c]: Swap line above with the blank line above

File: video/out/drm_atomic.c [hunk 16]
```diff
@@ -0,0 +1,229 @@
...
+            if (plane) {
+                if (drm_object_get_property(plane, "TYPE", &value) == -EINVAL) {
+                    mp_err(log, "Unable to retrieve type property from plane %d\n", j);
+                    goto fail;
+                } else {
+
```
Comment (rr-) [inline video/out/drm_atomic.c]: Superfluous newline

File: video/out/drm_atomic.c [hunk 17]
```diff
@@ -0,0 +1,229 @@
...
+                    if ((value == DRM_PLANE_TYPE_OVERLAY) && (layercount == overlay_id))
+                        ctx->overlay_plane = plane;
+                    else if (value == DRM_PLANE_TYPE_PRIMARY)
+                        ctx->primary_plane = plane;
+                    else {
```
Comment (rr-) [inline video/out/drm_atomic.c]: Stay consistent when using `{` `}`

File: video/out/opengl/context_drm_egl.c (hunk 5 above)
Comment (LongChair) [reply video/out/opengl/context_drm_egl.c]: that one doesn't show up here, but i also changed it.

File: libmpv/opengl_cb.h [hunk 18]
```diff
@@ -163,9 +175,13 @@ extern "C" {
...
- * glMPGetNativeDisplay("drm") should return a DRM FD casted to intptr_t (note
- * that a 0 FD is not supported - if this can happen in your case, you must
- * dup2() it to a non-0 FD).
+ * glMPGetNativeDisplay("drm") should return an array of void * containing
+ * 3 elements which are:
+ * 0 : pointer to the DRM fd (int). Note that a 0 FD is not supported - if this
+ *     can happen in your case, you must dup2() it to a non-0 FD).
+ * 1 : pointer to the crtc_id (int). This is the currently used crtc.
+ * 2 : pointer to the drmAtomicRequest that is being used for the renderloop.
+ *     This atomic request pointer should be usually created at every renderloop.
```
Comment (wm4) [inline libmpv/opengl_cb.h]: Should be renamed since the ABI changes. Also maybe it's time to make this a struct. Make sure the header builds on any OS (i.e. without DRM headers), which might require the use of opaque structs or even just `void*`.

File: libmpv/opengl_cb.h [hunk 19]
```diff
@@ -151,6 +151,18 @@ extern "C" {
...
+ *
+ * When using GL, sometimes GL rendering window is upscaled to display buffer.
+ * Typically with drm where GL framebuffer can be upscaled at later stage.
+ * In That case glMPGetNativeDisplay("opengl-cb") should return the GL window
+ * dimensions, that is 4 elements:
+ * 0 : left coordinates of window (usually 0)
+ * 1 : top coordinates of window (usually 0)
+ * 2 : width of GL window
+ * 3 : height of GL window
+ *
```
Comment (wm4) [inline libmpv/opengl_cb.h]: Better rename it to `opengl-cb-params` and make it a struct.

File: video/out/drm_atomic.c [hunk 20]
```diff
@@ -0,0 +1,229 @@
...
+   }
+
+   return -EINVAL;
+}
+
+struct drm_object * drm_object_create(struct mp_log *log, int fd, uint32_t object_id, uint32_t type)
+{
+    struct drm_object *obj = NULL;
+    obj = talloc_zero(NULL, struct drm_object);
+    if (obj) {
```
Comment (wm4) [inline video/out/drm_atomic.c]: Mos talloc functions (in mpv) actually just kill the process if allocation fails, so the check is unnecessary.

File: video/out/drm_atomic.c [hunk 21]
```diff
@@ -0,0 +1,229 @@
...
+       if (strcasecmp(name, object->props_info[i]->name) == 0) {
+           return drmModeAtomicAddProperty(request, object->id,
+                                           object->props_info[i]->prop_id, value);
+       }
+   }
+
+   return -EINVAL;
+}
+
+struct drm_object * drm_object_create(struct mp_log *log, int fd, uint32_t object_id, uint32_t type)
```
Comment (wm4) [inline video/out/drm_atomic.c]: Break all lines on 80 columns.

File: video/out/drm_atomic.c [hunk 22]
```diff
@@ -0,0 +1,229 @@
...
+    }
+}
+
+void drm_object_print_info(struct mp_log *log, struct drm_object *object)
+{
+    mp_err(log, "Object ID = %d (type = %x) has %d properties\n",
+           object->id, object->type, object->props->count_props);
+
+    for (int i = 0; i < object->props->count_props; i++)
+        mp_err(log, "\tProperty \"%s\" = %" PRId64 "\n",
```
Comment (wm4) [inline video/out/drm_atomic.c]: Don't use `\t`, just use spaces. Also in my opinion it's better to use `'` instead of `"` to make the string literal less terrible.

Also you could use `%lld` and cast the argument to `long long`, which might be slightly more readable and robust. (Same for all other places that do this.) Some consider this actually bad practice, but I think it's better because printf format arguments are so fragile.

File: video/out/drm_atomic.h [hunk 23]
```diff
@@ -0,0 +1,55 @@
...
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with mpv.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#ifndef DRMATOMIC_H
+#define DRMATOMIC_H
```
Comment (wm4) [inline video/out/drm_atomic.h]: You can use `#pragma once`, which is not standard C, but accepted in mpv. If you don't want to use that, add a `MP_` prefix to the define to avoid potential name clashes with upstream DRM.

File: video/out/drm_prime.c [hunk 24]
```diff
@@ -0,0 +1,86 @@
...
+
+int drm_prime_create_framebuffer(struct mp_log *log, int fd, AVDRMFrameDescriptor *descriptor, int width, int height,
+                                  struct  drm_prime_framebuffer *framebuffer)
+{
+    AVDRMLayerDescriptor *layer = NULL;
+    uint32_t pitches[4], offsets[4], handles[4];
+    int ret, layer_fd;
+
+    if (descriptor && descriptor->nb_layers) {
+        memset(framebuffer, 0, sizeof(*framebuffer));
```
Comment (wm4) [inline video/out/drm_prime.c]: `*framebuffer = (struct drm_prime_framebuffer){0};` is preferred.

File: video/out/opengl/context.c [hunk 25]
```diff
@@ -270,6 +270,9 @@ bool ra_gl_ctx_start_frame(struct ra_swapchain *sw, struct ra_fbo *out_fbo)
     struct priv *p = sw->priv;
     out_fbo->tex = p->wrapped_fb;
     out_fbo->flip = !p->params.flipped; // OpenGL FBs are normally flipped
+
+    if (p->params.start_frame)
+        p->params.start_frame(sw->ctx);
```
Comment (wm4) [inline video/out/opengl/context.c]: Would be nice if this could be avoided.

File: video/out/opengl/context_drm_egl.c [hunk 26]
```diff
@@ -119,7 +121,7 @@ static bool init_gbm(struct ra_ctx *ctx)
         p->gbm.device,
         p->kms->mode.hdisplay,
         p->kms->mode.vdisplay,
-        GBM_BO_FORMAT_XRGB8888,
+        GBM_FORMAT_XRGB8888,
```
Comment (wm4) [inline video/out/opengl/context_drm_egl.c]: This should probably be in a separate commit, together with the change below that changes the framebuffer depth.

File: video/out/opengl/context_drm_egl.c [hunk 27]
```diff
@@ -314,9 +339,9 @@ static bool drm_egl_init(struct ra_ctx *ctx)
         MP_WARN(ctx, "Failed to set up VT switcher. Terminal switching will be unavailable.\n");
     }
 
-    MP_VERBOSE(ctx, "Initializing KMS\n");
+    MP_VERBOSE(ctx->vo, "Initializing KMS\n");
```
Comment (wm4) [inline video/out/opengl/context_drm_egl.c]: Wut?

File: video/out/opengl/hwdec_drmprime_drm.c [hunk 28]
```diff
@@ -0,0 +1,255 @@
...
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with mpv.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include <assert.h>
+#include <errno.h>
+#include <libavutil/hwcontext_drm.h>
```
Comment (wm4) [inline video/out/opengl/hwdec_drmprime_drm.c]: Order external library dependencies between standard lib includes and mpv internal includes.

File: video/out/opengl/hwdec_drmprime_drm.c [hunk 29]
```diff
@@ -0,0 +1,255 @@
...
+
+    if (hw_image) {
+
+        // grab opengl-cb windowing info to eventually upscale the overlay
+        // as egl windows could be upscaled to primary plane.
+        int *data = gl ? mpgl_get_native_display(gl, "opengl-cb") : NULL;
+        if (data)
+            scale_dst_rect(hw, data[2], data[3], dst, &p->dst);
+        else
+            p->dst = *dst;
```
Comment (wm4) [inline video/out/opengl/hwdec_drmprime_drm.c]: Coding style (see CONTRIBUTE.md).

File: video/out/opengl/hwdec_drmprime_drm.c [hunk 30]
```diff
@@ -0,0 +1,255 @@
...
+        p->src = *src;
+
+        // grab drm interop info
+        void **drm_data;
+        drm_data =  gl ? mpgl_get_native_display(gl, "drm") : NULL;
```
Comment (wm4) [inline video/out/opengl/hwdec_drmprime_drm.c]: No reason to separate declaration and initialization.

File: video/out/opengl/hwdec_drmprime_drm.c [hunk 31]
```diff
@@ -0,0 +1,255 @@
...
+        drm_atomic_destroy_context(p->ctx);
+        p->ctx = NULL;
+    }
+}
+
+static int init(struct ra_hwdec *hw)
+{
+    struct priv *p = hw->priv;
+    int drm_overlay;
+    void **drm_data;
```
Comment (wm4) [inline video/out/opengl/hwdec_drmprime_drm.c]: Don't separate declaration and initialization (unless other reasons require it).

File: video/out/opengl/hwdec_drmprime_drm.c [hunk 32]
```diff
@@ -0,0 +1,255 @@
...
+
+    p->log = hw->log;
+
+    mp_read_option_raw(hw->global, "drm-overlay", &m_option_type_int, &drm_overlay);
```
Comment (wm4) [inline video/out/opengl/hwdec_drmprime_drm.c]: Avoid `mp_read_option_raw`, get the entire struct instead.

File: video/out/opengl/context.c (hunk 25 above)
Comment (LongChair) [reply video/out/opengl/context.c]: I don't see how to avoid this, the atmoic request must be share between on the opengl rendering loop and the hwdec. The atomic request needs to be created before each frame and put in the native struct to be shared. 
Do you have any better place in mind where I could do this ?

Comment (wm4) [reply video/out/opengl/context.c]: Why can't you do it in ra_swapchain_fns.start_frame?

File: video/out/opengl/hwdec_drmprime_drm.c (hunk 32 above)
Comment (LongChair) [reply video/out/opengl/hwdec_drmprime_drm.c]: I'm unsure how to do that. I have seen that the options are accessible thru `vo->opts` but o don't seem to have access to the vo from the hwdec interop. Could you clarify what you were expecting me to use please ?

File: libmpv/opengl_cb.h (hunk 19 above)
Comment (LongChair) [reply libmpv/opengl_cb.h]: Making a struct is fine, but i have a few questions about that. if we make a struct for this one, i suppose it makes sense to make also a struct for the drm one ? i suppose it should be `drm-params` then ?
Would these structs need to be expose in mpv headers when using it from third party app so that the interop native struct can be reused or should the thrid party app redefine it, and we only expose a `void *` and third party app would have to redefine it ? 
Either ways, what's the best place to define those structs ? `opengl__cb.h` ?

File: libmpv/opengl_cb.h (hunk 18 above)
Comment (LongChair) [reply libmpv/opengl_cb.h]: Agreed, `void *` should probably be fine as long as we document that properly in the headers.

File: video/out/opengl/context.c (hunk 25 above)
Comment (LongChair) [reply video/out/opengl/context.c]: Yeah I agree, didn't have noticed the swapchain. By the way looking at how this works, I'm wondering if it wouldn't be cleaner in the `context_drm_egl.c` to have different `ra_swapchain_fns` (one for drm legacy and one for drm atomic), set them accordingly in the `init`function depending on what's supported, and just have the code splitted in different functions. what would be your thought on this ?

File: video/out/opengl/context_drm_egl.c (hunk 26 above)
Comment (LongChair) [reply video/out/opengl/context_drm_egl.c]: OK i'll spit those changes into a separate commit.

File: libmpv/opengl_cb.h (hunk 19 above)
Comment (wm4) [reply libmpv/opengl_cb.h]: Yes. Preferable you can use opaque structs: `struct foo { struct something *p; };` is allowed and compiles even if `something` is never defined anywhere else in the header (not even by a recursive include). That's an opaque struct.

File: video/out/opengl/context.c (hunk 25 above)
Comment (wm4) [reply video/out/opengl/context.c]: Sounds ok too.

File: video/out/opengl/hwdec_drmprime_drm.c [hunk 33]
```diff
@@ -0,0 +1,255 @@
...
+}
+
+static int init(struct ra_hwdec *hw)
+{
+    struct priv *p = hw->priv;
+    int drm_overlay;
+
+    p->log = hw->log;
+
+    mp_read_option_raw(hw->global, "drm-overlay", &m_option_type_int, &drm_overlay);
```
Comment (wm4) [inline video/out/opengl/hwdec_drmprime_drm.c]: If the option value is not directly accessible somehow, use mp_get_config_group. video/out/gpu/spirv.c might be a good example of how to use it. (I regret starting using mp_read_option_raw anywhere, because it has even less type checking than the little the option system gives us.)

File: libmpv/opengl_cb.h [hunk 34]
```diff
@@ -151,6 +151,22 @@ extern "C" {
...
  *
+ * * Windowing system scaling
+ * ------------------------------------
+ *
+ * When using GL, sometimes GL rendering window is upscaled to display buffer.
+ * Typically with drm where GL framebuffer can be upscaled at later stage.
+ * In That case glMPGetNativeDisplay("opengl-cb-params") should return an
+ * mp_opengl_cb_params struct pointer :
+ */
+struct mp_opengl_cb_params {
```
Comment (wm4) [inline libmpv/opengl_cb.h]: Rename it to `opengl-cb-window-pos` and `mpv_opengl_cb_window_pos`, and mention that this is an obscure mechanism for some hardware overlays.

File: libmpv/opengl_cb.h [hunk 35]
```diff
@@ -163,10 +179,22 @@ extern "C" {
  *
  * glMPGetNativeDisplay("wl") should return a Wayland "struct wl_display *".
  *
- * glMPGetNativeDisplay("drm") should return a DRM FD casted to intptr_t (note
- * that a 0 FD is not supported - if this can happen in your case, you must
- * dup2() it to a non-0 FD).
- *
+ * glMPGetNativeDisplay("drm-params") should return an mp_drm_params structure pointer:
+ */
+struct mp_drm_params {
```
Comment (wm4) [inline libmpv/opengl_cb.h]: Rename to `mpv_opengl_cb_drm_params`. Identifiers starting with `mp_` are always private to mpv.

File: libmpv/opengl_cb.h [hunk 36]
```diff
@@ -163,10 +179,22 @@ extern "C" {
...
+    // DRM fd (int). Note that a 0 FD is not supported - if this
+    // can happen in your case, you must dup2() it to a non-0 FD).
+    int fd;
+
+    // currently used crtc id
+    int crtc_id;
+
+    // pointer to the drmAtomicRequest that is being used for the renderloop.
+    // This atomic request pointer should be usually created at every renderloop.
+    void *atomic_request;
```
Comment (wm4) [inline libmpv/opengl_cb.h]: It's actually `drmModeAtomicReq` not `drmAtomicRequest`. Also I think you can change this to `struct _drmModeAtomicReq *atomic_request;`. I don't know why the header authors added this dumb `_` prefix - maybe it's meant to be private, but this is not a C convention, it's a violation of the C standard. So whatever.

File: libmpv/opengl_cb.h [hunk 37]
```diff
@@ -163,10 +179,22 @@ extern "C" {
...
- * glMPGetNativeDisplay("drm") should return a DRM FD casted to intptr_t (note
- * that a 0 FD is not supported - if this can happen in your case, you must
- * dup2() it to a non-0 FD).
- *
+ * glMPGetNativeDisplay("drm-params") should return an mp_drm_params structure pointer:
+ */
+struct mp_drm_params {
+    // DRM fd (int). Note that a 0 FD is not supported - if this
+    // can happen in your case, you must dup2() it to a non-0 FD).
+    int fd;
```
Comment (wm4) [inline libmpv/opengl_cb.h]: There's no need for that 0 special case. It was sort of needed with the old openglcb/drm API, because NULL was the default return value for "not present". Define -1 as invalid/not-present/unused value instead.

File: libmpv/opengl_cb.h [hunk 38]
```diff
@@ -201,6 +229,8 @@ extern "C" {
  * deallocated by the API user.
  */
 
+
+
```
Comment (wm4) [inline libmpv/opengl_cb.h]: Stray space.

File: video/out/drm_atomic.c [hunk 39]
```diff
@@ -0,0 +1,244 @@
...
+
+int drm_object_create_properties(struct mp_log *log, int fd,
+                                 struct drm_object *object)
+{
+    object->props = drmModeObjectGetProperties(fd, object->id, object->type);
+    if (object->props) {
+        object->props_info = talloc_zero_size(NULL, object->props->count_props
+                                              * sizeof(object->props_info));
+        if (object->props_info) {
+            for (int i=0; i < object->props->count_props; i++)
```
Comment (wm4) [inline video/out/drm_atomic.c]: Missing space around `=`.

File: video/out/drm_atomic.c (hunk 14 above)
Comment (wm4) [reply video/out/drm_atomic.c]: (Still not changed.)

File: video/out/drm_atomic.c (hunk 17 above)
Comment (wm4) [reply video/out/drm_atomic.c]: (Still not changed.)

File: video/out/drm_prime.c [hunk 40]
```diff
@@ -0,0 +1,86 @@
...
+        layer = &descriptor->layers[0];
+
+        for (int plane = 0; plane < AV_DRM_MAX_PLANES; plane++) {
+            layer_fd = framebuffer->gem_handles[layer->planes[plane].object_index];
+            if (layer_fd && layer->planes[plane].pitch) {
+                pitches[plane] = layer->planes[plane].pitch;
+                offsets[plane] = layer->planes[plane].offset;
+                handles[plane] = layer_fd;
+            }
+            else {
```
Comment (wm4) [inline video/out/drm_prime.c]: Coding style.

File: video/out/drm_prime.c [hunk 41]
```diff
@@ -0,0 +1,86 @@
...
+   memset(framebuffer, 0, sizeof(*framebuffer));
+   return -1;
+}
+
+void drm_prime_destroy_framebuffer(struct mp_log *log, int fd, struct  drm_prime_framebuffer *framebuffer)
+{
+    if (framebuffer->fb_id)
+        drmModeRmFB(fd, framebuffer->fb_id);
+
+    for (int i=0; i < AV_DRM_MAX_PLANES; i++) {
```
Comment (wm4) [inline video/out/drm_prime.c]: Space around `=`.

File: video/out/opengl/context_drm_egl.c [hunk 42]
```diff
@@ -360,10 +392,15 @@ static bool drm_egl_init(struct ra_ctx *ctx)
...
 
+    p->drm_params.fd = p->kms->fd;
+    p->drm_params.crtc_id = p->kms->crtc_id;
+    p->drm_params.atomic_request = p->kms->atomic_context->request;
     struct ra_gl_ctx_params params = {
         .swap_buffers = drm_egl_swap_buffers,
-        .native_display_type = "drm",
-        .native_display = (void *)(intptr_t)p->kms->fd,
+        .native_display_type = "drm-params",
+        .native_display = (void *)&p->drm_params,
```
Comment (wm4) [inline video/out/opengl/context_drm_egl.c]: Since this is already a pointer, the void* cast is not needed.

File: video/out/opengl/hwdec_drmprime_drm.c [hunk 43]
```diff
@@ -0,0 +1,257 @@
...
+                         struct mp_rect *src, struct mp_rect *dst, bool newframe)
+{
+    struct priv *p = hw->priv;
+    GL *gl = ra_gl_get(hw->ra);
+    AVDRMFrameDescriptor *desc = NULL;
+    drmModeAtomicReq *request = NULL;
+    struct drm_frame next_frame;
+    int ret;
+
+    memset(&next_frame, 0, sizeof(struct drm_frame));
```
Comment (wm4) [inline video/out/opengl/hwdec_drmprime_drm.c]: Use assignment + initializer instead of memset.

**Comment by wm4** [comment]
Merged. Added some forgotten things in 2c006876418b12c and 45930022225 myself.
